File: DiscoDruid3/app/src/main/java/com/andruid/magic/discodruid/viewmodel/AlbumViewModel.java
package com.andruid.magic.discodruid.viewmodel;

import android.app.Application;
import android.support.v4.media.MediaBrowserCompat;

import com.andruid.magic.discodruid.data.Constants;
import com.andruid.magic.discodruid.datasourcefactory.AlbumDataSourceFactory;
import com.andruid.magic.discodruid.model.Album;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.paging.LivePagedListBuilder;
import androidx.paging.PagedList;

public class AlbumViewModel extends AndroidViewModel {
    private Application application;

    public AlbumViewModel(@NonNull Application application) {
        super(application);
        this.application = application;
    }

    public LiveData<PagedList<Album>> getAlbums(MediaBrowserCompat mediaBrowserCompat){
        PagedList.Config config = new PagedList.Config.Builder()
                .setEnablePlaceholders(false)
                .setPageSize(Constants.PAGE_SIZE)
                .build();
        return new LivePagedListBuilder<>(new AlbumDataSourceFactory(mediaBrowserCompat,application.getApplicationContext()),config).build();
    }
}

File: DiscoDruid/app/src/main/java/com/andruid/magic/discodruid/ui/viewmodel/AlbumViewModel.kt
package com.andruid.magic.discodruid.ui.viewmodel

import android.os.Bundle
import android.support.v4.media.MediaBrowserCompat
import androidx.lifecycle.ViewModel
import androidx.lifecycle.asLiveData
import androidx.lifecycle.liveData
import androidx.lifecycle.viewModelScope
import androidx.paging.Pager
import androidx.paging.PagingConfig
import androidx.paging.cachedIn
import com.andruid.magic.discodruid.data.PAGE_SIZE
import com.andruid.magic.discodruid.paging.AlbumsPagingSource
import kotlinx.coroutines.cancel

class AlbumViewModel(
    private val mediaBrowserCompat: MediaBrowserCompat,
    private val options: Bundle? = null
) : ViewModel() {
    val albumsLiveData = liveData {
        val config = PagingConfig(PAGE_SIZE)
        val pager = Pager(config) {
            AlbumsPagingSource(mediaBrowserCompat, options)
        }

        emitSource(
            pager.flow.cachedIn(viewModelScope)
                .asLiveData()
        )
    }

    override fun onCleared() {
        super.onCleared()
        viewModelScope.cancel()
    }
}